spark/src/main/scala/com/gu/tableversions/spark/filesystem/ProxyFileSystem.scala

```diff
@@ -0,0 +1,103 @@
+package com.gu.tableversions.spark.filesystem
+
+import java.net.URI
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+import org.apache.hadoop.fs.permission.FsPermission
+import org.apache.hadoop.util.Progressable
+
+/**
+  * A ProxyFileSystem wraps the behaviour of some other ("underlying" or "base") FileSystem,
+  * enabling transformations of `Path`s provided to the base `FileSystem` through the use of the
+  * `PathMapper`. Operations on the `ProxyFileSystem` ultimately delegate to the base filesystem,
+  * but first translate the paths provided to the `ProxyFileSystem` into a format suitable for
+  * that base filesystem.
+  *
+  * `Path` values returned by calls to the base filesystem are translated in the opposite
+  * direction by the `PathMapper`, making them suitable for use in future calls to the ProxyFileSystem.
+  */
+abstract class ProxyFileSystem extends FileSystem {
+
+  // The underlying FileSystem we delegate to. Paths in calls to this FileSystem
+  // should be translated with the PathMapper.
+  @volatile protected var baseFs: FileSystem = _
+
+  // The root URI for the underlying FileSystem
+  @volatile protected var baseUri: URI = _
+
+  // A utility to map between Paths for the proxy and underlying FileSystems.
+  @volatile protected var pathMapper: PathMapper = _
+
+  /** Call from concrete implementation to initialise the base class. */
+  protected def initialiseProxyFileSystem(_baseUri: URI, _pathMapper: PathMapper, hadoopConfig: Configuration): Unit = {
+    baseUri = _baseUri
+    pathMapper = _pathMapper
+
+    baseFs = FileSystem.get(baseUri, hadoopConfig)
+    baseFs.initialize(baseUri, hadoopConfig)
+  }
+
+  override def getUri: URI =
+    new URI(s"${VersionedFileSystem.scheme}:${baseUri.getSchemeSpecificPart}")
+
+  override def open(f: Path, bufferSize: Int): FSDataInputStream =
+    baseFs.open(pathMapper.forUnderlying(f), bufferSize)
+
+  override def create(
+      f: Path,
+      permission: FsPermission,
+      overwrite: Boolean,
+      bufferSize: Int,
+      replication: Short,
+      blockSize: Long,
+      progress: Progressable): FSDataOutputStream =
+    baseFs.create(pathMapper.forUnderlying(f), permission, overwrite, bufferSize, replication, blockSize, progress)
```

Comment:
Minor. The `forUnderlying` and `toUnderlying` would read a tad nicer if thy pimped `Path`, ie `f.forUnderlying` ?

Reply:
I had a look at this but it's a bit messy as the path mapper would have to be an implicit argument then, and it's a `var` which can't be implicit...

As this is just used internally in this class I think it's reasonable to leave it as it is.